client/opentelemetry.go

```diff
@@ -0,0 +1,111 @@
+package client
+
+import (
+	"net/http"
+
+	"github.com/go-openapi/runtime"
+	"github.com/go-openapi/strfmt"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type openTelemetryTransport struct {
+	transport runtime.ClientTransport
+	host      string
+	opts      []trace.SpanStartOption
+}
+
+func newOpenTelemetryTransport(transport runtime.ClientTransport, host string, opts []trace.SpanStartOption) runtime.ClientTransport {
+	return &openTelemetryTransport{
+		transport: transport,
+		host:      host,
+		opts:      opts,
+	}
+}
+
+func (t *openTelemetryTransport) Submit(op *runtime.ClientOperation) (interface{}, error) {
+	if op.Context == nil {
+		return t.transport.Submit(op)
+	}
+
+	params := op.Params
+	reader := op.Reader
+
+	var span trace.Span
+	defer func() {
+		if span != nil {
+			span.End()
+		}
+	}()
+
+	op.Params = runtime.ClientRequestWriterFunc(func(req runtime.ClientRequest, reg strfmt.Registry) error {
+		span = createOpenTelemetryClientSpan(op, req.GetHeaderParams(), t.host, t.opts)
+		return params.WriteToRequest(req, reg)
+	})
+
+	op.Reader = runtime.ClientResponseReaderFunc(func(response runtime.ClientResponse, consumer runtime.Consumer) (interface{}, error) {
+		if span != nil {
+			statusCode := response.Code()
+			span.SetAttributes(attribute.Int(string(semconv.HTTPStatusCodeKey), statusCode))
+			span.SetStatus(semconv.SpanStatusFromHTTPStatusCodeAndSpanKind(statusCode, trace.SpanKindClient))
+		}
+
+		return reader.ReadResponse(response, consumer)
+	})
+
+	submit, err := t.transport.Submit(op)
+	if err != nil && span != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, "")
+	}
+
+	return submit, err
+}
+
+func createOpenTelemetryClientSpan(op *runtime.ClientOperation, _ http.Header, host string, opts []trace.SpanStartOption) trace.Span {
+	ctx := op.Context
+	span := trace.SpanFromContext(ctx)
+
+	if span != nil {
+		// TODO: Can we get the version number for use with trace.WithInstrumentationVersion?
+		tracer := otel.GetTracerProvider().Tracer("")
+
+		ctx, span = tracer.Start(ctx, operationName(op), opts...)
+		op.Context = ctx
+
+		span.SetAttributes(
+			attribute.String("net.peer.name", host),
+			// attribute.String("net.peer.port", ""),
+			attribute.String(string(semconv.HTTPRouteKey), op.PathPattern),
+			attribute.String(string(semconv.HTTPMethodKey), op.Method),
+			attribute.String("span.kind", trace.SpanKindClient.String()),
+			//TODO: Figure out the best way to get the scheme?
+			attribute.String("http.scheme", op.Schemes[0]),
```

Comment:
Hi @casualjim, can you offer any pointers on the best way to get the scheme? I don't appear to have access to the underlying `http.Request` object to check the existence of the `TLS` field (which is one common way to work out the scheme)